File: python/tests/threading_test.py
import time, json, gc
from concurrent.futures import ThreadPoolExecutor

class ThreadingTestTest:
    name = "Threading / Concurrency"

    def __init__(self, threads=8, work_size=2_000_000, runs=3):
        self.threads = threads
        self.work_size = work_size
        self.runs = runs

    def cpu_bound_work(self, n):
        s = 0
        for i in range(n):
            s += (i * i) % 97
        return s

    def thread_task(self):
        start = time.perf_counter()
        self.cpu_bound_work(self.work_size // self.threads)
        end = time.perf_counter()
        return end - start

    def run_once(self):
        gc.collect()

        t0 = time.perf_counter()

        with ThreadPoolExecutor(max_workers=self.threads) as pool:
            results = list(pool.map(lambda _: self.thread_task(), range(self.threads)))

        t1 = time.perf_counter()

        total_time = t1 - t0
        avg_thread_time = sum(results) / self.threads
        ops_per_s = self.work_size / total_time if total_time > 0 else float("inf")

        return {
            "total_time_s": total_time,
            "avg_thread_time_s": avg_thread_time,
            "ops_per_s": ops_per_s
        }

    def run(self, runs=None, iterations=None):
        if runs is None:
            runs = self.runs

        results = [self.run_once() for _ in range(runs)]
        times = sorted(r["total_time_s"] for r in results)
        ops   = sorted(r["ops_per_s"] for r in results)
        avg_t = sorted(r["avg_thread_time_s"] for r in results)

        mid = len(times) // 2

        return {
            "name": self.name,
            "runs": runs,
            "threads": self.threads,
            "median_total_time_s": times[mid],
            "median_avg_thread_time_s": avg_t[mid],
            "median_ops_per_s": ops[mid],
            "raw": results
        }

if __name__ == "__main__":
    result = ThreadingTestTest().run()
    with open("results_python_threading.json", "w") as f:
        json.dump(result, f, indent=2)
    print(json.dumps(result, indent=2))

File: ruby/tests/threading_test.rb
require "json"
require "time"

class ThreadingTestTest
  attr_reader :name

  def initialize(threads = 8, work_size = 2_000_000)
    @name = "Threading / Concurrency"
    @threads = threads
    @work_size = work_size
  end

  def cpu_bound_work(n)
    s = 0
    n.times do |i|
      s += (i * i) % 97
    end
    s
  end

  def single_run
    results = Array.new(@threads, 0.0)
    threads = []

    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)

    @threads.times do |i|
      threads << Thread.new do
        t_start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        cpu_bound_work(@work_size / @threads)
        t_end = Process.clock_gettime(Process::CLOCK_MONOTONIC)
        results[i] = t_end - t_start
      end
    end

    threads.each(&:join)

    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    total_time = t1 - t0
    avg_thread_time = results.sum / @threads
    ops_per_s = @work_size / (total_time > 0 ? total_time : 1e-9)

    {
      "total_time_s" => total_time,
      "avg_thread_time_s" => avg_thread_time,
      "ops_per_s" => ops_per_s
    }
  end

  def run(runs = 3, _initialize = nil)
    results = Array.new(runs) { single_run }
    valid_results = results.reject { |r| r.nil? || r.empty? }

    if valid_results.empty?
      return { "name" => @name, "error" => "All runs failed" }
    end

    total_times = valid_results.map { |r| r["total_time_s"] }.sort
    avg_thread_times = valid_results.map { |r| r["avg_thread_time_s"] }.sort
    ops_per_ss = valid_results.map { |r| r["ops_per_s"] }.sort
    median_idx = valid_results.size / 2

    {
      "name" => @name,
      "threads" => @threads,
      "runs" => runs,
      "median_total_time_s" => total_times[median_idx],
      "median_avg_thread_time_s" => avg_thread_times[median_idx],
      "median_ops_per_s" => ops_per_ss[median_idx],
      "raw" => results
    }
  rescue Interrupt
    { "name" => @name, "error" => "Interrupted during test" }
  end
end

if __FILE__ == $0
  test = ThreadingTestTest.new
  result = test.run
  timestamp = Time.now.strftime('%Y%m%d_%H%M%S')
  filename = "results_threading_ruby_#{timestamp}.json"
  File.write(filename, JSON.pretty_generate(result))
  puts "Results saved to #{filename}"
  puts JSON.pretty_generate(result)
end